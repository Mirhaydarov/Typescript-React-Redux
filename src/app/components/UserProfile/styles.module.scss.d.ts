export const wrap: string;
export const spinner: string;
export const profile: string;
export const title: string;
export const avatar: string;
export const listContainer: string;
export const listTitle: string;
export const listUser: string;
export const listCompany: string;
export const listAddress: string;
export const listItem: string;
export const listItemDesc: string;
export const baseButton: string;
export const button: string;
