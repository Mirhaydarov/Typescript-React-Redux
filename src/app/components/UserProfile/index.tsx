// Core
import React, { FC, ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';

// Components
import { ErrorIndicator } from '../ErrorIndicator';

// Reducers
import { AppState } from '../../init/rootReducer';
import { UserListState } from '../../bus/UserList/reducer';

// Types
import { UserList } from '../../bus/UserList/types';

// Styles
import styles from './styles.module.scss';

type RouteParams = {
  userName: string;
};

type UserProfileProps = RouteComponentProps<RouteParams>;

export const currentUser = (
  user: UserList | undefined,
  matchParams: string
): UserList | boolean => {
  if (user === undefined) return false;
  return user.name.replace(/ /g, '-').toLowerCase() === matchParams;
};

export const UserProfile: FC<UserProfileProps> = (
  props: UserProfileProps
): ReactElement => {
  const { data } = useSelector<AppState, UserListState>(
    (state) => state.userList
  );

  const {
    match: { params },
  } = props;
  const { userName } = params;

  const user = data.find((users) => currentUser(users, userName));

  const renderProfile = (): ReactElement | boolean => {
    if (user) {
      const {
        name,
        email,
        phone,
        image,
        address: { city, street },
        company: { name: companyName, catchPhrase },
      } = user;

      return (
        <>
          <h2 className={styles.title}>
            Profile:&nbsp;
            {name}
          </h2>
          <div className={styles.listContainer}>
            <img className={styles.avatar} src={image} alt={name} />
            <div>
              <ul className={styles.listUser}>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>Name:</span>
                  &nbsp;
                  {name}
                </li>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>Email:</span>
                  &nbsp;
                  {email}
                </li>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>Phone:</span>
                  &nbsp;
                  {phone}
                </li>
              </ul>
            </div>
            <div>
              <h3 className={styles.listTitle}>Company</h3>
              <ul className={styles.listCompany}>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>Name:</span>
                  &nbsp;
                  {companyName}
                </li>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>
                    Catch phrase:
                  </span>
                  &nbsp;
                  {catchPhrase}
                </li>
              </ul>
            </div>
            <div>
              <h3 className={styles.listTitle}>Address</h3>
              <ul className={styles.listAddress}>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>City:</span>
                  &nbsp;
                  {city}
                </li>
                <li className={styles.listItem}>
                  <span className={styles.listItemDesc}>Street:</span>
                  &nbsp;
                  {street}
                </li>
              </ul>
            </div>
          </div>
        </>
      );
    }
    return false;
  };

  const UserProfileJSX = user ? renderProfile() : <ErrorIndicator />;

  return (
    <section className={styles.profile}>
      <div className={styles.wrap}>{UserProfileJSX}</div>
    </section>
  );
};
