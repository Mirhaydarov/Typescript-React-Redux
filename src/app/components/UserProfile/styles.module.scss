$primary--color: #333333;
$border--color: #d6d6d6;
$body-bgc--color: #dddddd;
$body-bgc--hover-color: #89a8ff;

$head--font-size: 1.5rem;
$body--font-size: 1.2rem;
$body--image-max-width: 100%;
$body--image-width: 7rem;
$body--image-height: 7rem;
$body--image-radius: 50%;

$head--padding: 0.5rem;
$body--padding: 1rem;

@mixin listTemplate($areaValue) {
  grid-area: $areaValue;
  list-style-type: none;
  font-size: 1.2rem;
}

.wrap {
  composes: wrap from '../../../root-styles.scss';
}

.spinner {
  text-align: center;
  margin: 0 auto;
}

.profile {
  padding-top: 50px;
}

.title {
  font-size: 2.5rem;
  text-align: center;
  padding-bottom: 50px;
}

.avatar {
  grid-area: profile-avatar;
  border-radius: 50%;
  width: 150px;
  height: 150px;
  object-fit: cover;
  margin: 0 auto;
}

.list-container {
  display: grid;
  grid-template-areas: 
      "profile-avatar list-user" 
      "list-company list-address";
  justify-content: center;
  grid-gap: 20px;
  align-items: center;
}

.list-title {
  font-size: 2rem;
  text-align: center;
}

.list-user {
  @include listTemplate(list-user);
}

.list-company {
  @include listTemplate(list-company);
}

.list-address {
  @include listTemplate(list-address);
}

.list-item {
  display: grid;
  grid-template-columns: max-content 1fr;
}

.list-item-desc {
  font-weight: 700;
}

.baseButton {
  border: 2px solid darkgray;
  background-color: gray;
}

.button {
  composes: button from "../../../root-styles.scss";
}