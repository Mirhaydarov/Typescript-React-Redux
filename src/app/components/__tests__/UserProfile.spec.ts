// Core
import { match } from 'react-router-dom';
import { createMemoryHistory } from 'history';

// Components
import { UserProfile, currentUser } from '../UserProfile';

// Instruments
import {
  makeMountRender,
  reduxify,
  store,
} from '../../../test-utils';

const history = createMemoryHistory();

const match: match<{ userName: string }> = {
  isExact: false,
  path: '',
  url: '',
  params: { userName: 'wrong-name' },
};

describe('<UserProfile />', () => {
  it("Should render an <ErrorIndicator /> if didn't find a match in data", () => {
    const routeComponentPropsMock = {
      location: {} as unknown,
      history,
      match,
    };
    const wrapper = makeMountRender(
      reduxify(UserProfile, store, routeComponentPropsMock)
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('currentUser return true', () => {
    const user = {
      id: 1,
      name: 'Leanne Graham',
      email: 'Sincere@april.biz',
      phone: '1-770-736-8031 x56442',
      image:
        'https://cdn.pixabay.com/user/2020/10/03/07-51-18-705_250x250.jpg',
      address: {
        street: 'Kulas Light',
        city: 'Gwenborough',
      },
      company: {
        name: 'Romaguera-Crona',
        catchPhrase: 'Multi-layered client-server neural-net',
      },
    };
    const userName = 'leanne-graham';

    const isUser = currentUser(user, userName);

    expect(isUser).toEqual(true);
  });

  it('currentUser return false', () => {
    const userName = 'leanne-graham';

    const isUser = currentUser(undefined, userName);

    expect(isUser).toEqual(false);
  });
});
