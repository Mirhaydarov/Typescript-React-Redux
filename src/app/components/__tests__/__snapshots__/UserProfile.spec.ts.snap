// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile /> Should render an <ErrorIndicator /> if didn't find a match in data 1`] = `
<section class="profile">
  <div class="wrap">
    <div class="errorIndicator">
      <h2>
        Error
      </h2>
      <p>
        Something has gone terribly wrong
      </p>
      <p>
        (but we already try to fix it)
      </p>
    </div>
  </div>
</section>
`;
